Lotto refactoring & 5단계 - 로또(수동)

SHINMJ opened the pull request:
리뷰 정말 감사합니다.
리뷰어님의 코멘트가 정말 많이 도움이 되었습니다.
객체지향적 TDD를 할 수 있도록 도와 주셔서 감사합니다.

새로운 것도 많이 알게되고(e.g. 디미터법칙..) 객체지향에서 가장 중요한 객체에게 물어보는 법에 대해 깨달았습니다.
그리고 각 클래스와 메서드의 책임과 역할을 명확이 구분하는 법을 배웠습니다.

이번엔 라이브 강의와 리뷰어님의 리뷰를 바탕으로 전체적인 코드 리팩토링과 미션 5단계도 마무리하였습니다!!

이번 코드도 잘 부탁드립니다.

SHINMJ committed:
docs: README.md update

- 로또 리팩토링 + 로또(수동) 기능 구현 목록 추가

SHINMJ committed:
refactor: 로또 미션 TDD 방식으로 다시 개발

- TDD 방식으로 다시 개발하기 위해 기존 소스 패키지 변경
- 사용자 input 처리 구현

SHINMJ committed:
refactor: 구입금액 클래스

- 구입금액 Integer 포장 클래스 추가
- 구입 금액 클래스 변경

SHINMJ committed:
refactor: 로또 번호 클래스 추가

- Integer 포장 클래스
- 1~45 사이의 양의 숫자 체크

SHINMJ committed:
refactor: 로또 클래스 생성

- 로또 번호 클래스를 리스트로 관리하는 일급 컬렉션
- 중복되지 않는 로또 번호 6개 생성

SHINMJ committed:
refactor: 로또 리스트 클래스 생성

- 로또 리스트 클래스 생성
- 구입 수량 만큼 로또 클래스 생성

SHINMJ committed:
feat,refactor: 수동 로또 생성

- 사용자 입력 로또 클래스 생성자 추가
- validation 추가
- 에러 메세지 추가

SHINMJ committed:
feat,refactor: 수동 로또 목록 생성

SHINMJ committed:
refactor: 보너스 볼 클래스

- 보너스볼 클래스 생성

SHINMJ committed:
refactor: Rank 클래스 생성

SHINMJ committed:
feat,refactor: 수동갯수클래스추가, 당첨관리 클래스 리팩토링

- Rank 테스트 케이스 추가
- 수동 갯수 클래스 추가
- 당첨관리 클래스 수정

SHINMJ committed:
feat,refactor: TDD 로 리팩토링 & 로또(수동)

- 전체 클래스 리팩토링 완료
- 로또 수동 기능 완료

SHINMJ committed:
refactor,style: 기존 레거시 코드 삭제, 코드 포맷팅

SHINMJ committed:
refactor: 로또 리스트 클래스 toString 변경

- toString 시 newline 추가

jhhj424 commented on src/main/java/common/CommonUtils.java:
```diff
@@ -0,0 +1,23 @@
+import exception.BusinessException;
+import exception.ErrorMessages;
+
+public class CommonUtils {
```
static 필드와 메서드만 가지는 클래스의 경우 인스턴스화가 필요하지 않기 때문에 생성자를 막는것이 좋을 것 같아요~

이펙티브 자바 Item 4번 내용 참고하시면 도움이 될 것 같습니다.
다른 쪽도 적용할 수 있는곳이 있는지 확인 부탁드려요~
참고: https://madplay.github.io/post/enforce-noninstantiability-with-private-constructor

jhhj424 commented on src/main/java/common/Constants.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.regex.Pattern;
+
+public class Constants {
+	public static final int PER_PRICE = 1000;
```
가독성을 위해 습관적으로 단위 언더바(_) 를 사용해봐도 좋을 것 같습니다 ~
```suggestion
	public static final int PER_PRICE = 1_000;
```

jhhj424 commented on src/main/java/exception/BusinessException.java:
```diff
@@ -0,0 +1,16 @@
+package exception;
+
+public class BusinessException extends RuntimeException {
```
Custom Exception 정의 💯

jhhj424 commented on src/main/java/exception/ErrorMessages.java:
```diff
@@ -0,0 +1,23 @@
+	INPUT_LOTTO_RANGE_NOT_VALID("1~45 까지의 숫자만 입력 가능합니다."),
+	INPUT_NUMBER_COMMA_FORMAT_NOT_VALID("숫자와 ','만 입력 가능합니다."),
+	INPUT_NUMBER_LENGTH_NOT_VALID("6개의 숫자를 입력해야 합니다."),
+	INPUT_NUMBER_DUPLICATE("중복된 숫자가 존재합니다."),
```
```suggestion
	INPUT_NUMBER_DUPLICATE("중복된 숫자가 존재합니다.");
```

jhhj424 commented on src/main/java/lotto/BonusBall.java:
```diff
@@ -1,40 +1,33 @@
-		} catch (NumberFormatException e) {
-			throw new IllegalArgumentException(Messages.INPUT_NUMBER_FORMAT_NOT_VALID.getValues());
-		}
+	public boolean matched(Lotto lotto) {
```
클래스 구현할 때 선언 순서를 이렇게 보통 사용합니다.

``` java
class Class {
    클래스 변수
    인스턴스 변수
    생성자
    팩토리 메소드
    메소드
    기본 메소드 (equals, hashCode, toString)
}
```

해당 메서드는 equals, hashCode 보다 상위로 이동시켜주는게 좋을 것 같습니다 👍

jhhj424 commented on src/test/java/common/MessagesTest.java:
```diff
@@ -0,0 +1,18 @@
+
+	@Test
+	@DisplayName("{arg} 있는 메시지 파라미터로 formatting하여 반환한다.")
+	public void messages_replace_format_value() {
```
테스트 코드도 자바의 메서드 네이밍 규칙인 변형된 카멜 케이스를 적용해주면 좋을 것 같습니다 :)

참고: https://donologue.tistory.com/168

jhhj424 commented on src/test/java/common/MessagesTest.java:
```diff
@@ -0,0 +1,18 @@
+	public void messages_replace_format_value() {
+		assertThat(Messages.BOUGHT_OF_FORMAT.getValues(new String[] {"3", "11"}))
+			.isEqualTo("수동으로 3장, 자동으로 11개를 구매했습니다.");
+		assertThat(Messages.RESULT_FORMAT.getValues(new String[] {"2.11"})).isEqualTo("총 수익률은 2.11입니다.");
```
하나의 테스트에서 두개 이상의 검증을 해야 할 경우 assertAll 로 묶어줄 수 있습니다!
관련해서는 assertAll 에 대한 내용을 찾아보시면 좋을 것 같아요 ~
하나씩 검증하는것과 동작에 대한 차이가 조금 있습니다 !

다른곳도 확인해주세요~

참고: https://sas-study.tistory.com/316, https://escapefromcoding.tistory.com/358?category=1184809

jhhj424 commented on src/test/java/investment/InvestmentTest.java:
```diff
@@ -0,0 +1,18 @@
+public class InvestmentTest {
+
+	@Test
+	public void 금액_2000원_입력_수량_2() {
```
@DisplayName 으로 테스트코드의 의도를 더 명확히 작성해보는 습관을 챙겨보면 좋을 것 같아요 👍 

다른곳도 확인해주세요~

jhhj424 commented on src/test/java/lotto/LottoTest.java:
```diff
@@ -2,17 +2,80 @@
+		String input = "1,2,3,4,5,6";
+		Lotto lotto = new Lotto(input);
+		assertThat(lotto.size()).isEqualTo(6);
+		System.out.println(lotto);
```
System.out.println 는 사용하지 않아도 될 듯 합니다!
삭제가 좋겠어요 ~

jhhj424 commented on README.md:
```diff
@@ -58,6 +58,65 @@ LottoMain 클래스의 main 메서드를 활용한다.
 
 
+## 로또 리팩토링 + 로또 (수동)
+### 기능 구현 목록
```
꼼꼼하게 잘 작성해주셨네요 👍

jhhj424 commented on src/main/java/lotto/Lotto.java:
```diff
@@ -1,28 +1,90 @@
 
-	private final List<LottoNumber> allNumbers =
-		IntStream.range(START_NUMBER, END_NUMBER).mapToObj(LottoNumber::new).collect(Collectors.toList());
+	private List<LottoNumber> lotto;
```
현재 Lotto 클래스에서 두개의 `List<LottoNumber>` 변수를 가지고 있는데 혹시 하나로 합쳐볼 수 없을까요?

조금만 더 고민해봐도 좋을 것 같아요 :)

일급 컬렉션과 관련해서는 해당 글 참고해보시면 좋을 것 같습니다!
참고: https://jojoldu.tistory.com/412

jhhj424 reviewed (changes requested):
안녕하세요 명진님! 🙇🏻‍♂️
로또 미션 마지막 단계 구현 잘 해주셨네요~ 👍 
계속 더 많은 고민과 함께 좋은 코드를 작성해주시려는게 느껴져서 저도 너무 감사합니다 :)
마지막 단계인만큼 조금의 욕심이 생겨서 Request changes 한번 드리겠습니다.
참고해볼만한 코멘트 조금 남겼으니 확인해주시고 마무리 하면 더 의미있는 미션이 될 것 같네요 😁

SHINMJ committed:
refactor: 코드리뷰를 통한 리팩토링

- static 필드와 메서드만 가지는 클래스 생성자 숨기기
- 숫자는 '_' 포함
- 클래스 구현 순서
- 자바 네이밍규칙 적용
- assertAll 사용
- 불필효한 System.out.println 삭제
- 일급컬렉션 정리

SHINMJ committed:
test,refactor: 테스트 코드 리팩토링

- 불필요한 throws 삭제
- 자바 네이밍 컨벤션 적용

jhhj424 reviewed (approved):
안녕하세요 명진님~ 🙇🏻‍♂️
피드백 반영 꼼꼼하게 잘 해주셨네요 💯 
길다면 길고 짧다면 짧은 기간인데 미션 단계마다 더 많은 고민을 거쳐서 코드를 작성하시는게 느껴져서 저도 덩달아서 열심히 리뷰하게 되었던 것 같습니다 !
저랑 함께하는 미션은 여기서 끝이지만 언제든 고민 나누고 싶으시면 DM 주세요 ~
남은 기간도 지금처럼 화이팅입니다 👊
고생하셨습니다! :)
